Fix encoding error on Windows with Python 3.6

JGoutin opened the pull request:
Fix for #818 

And tried to add Python 3.6 to AppVeyor test.

aragilar commented:
There are a few things missing from the 3.6 CI support, I've created a separate pull request #838 for that.

The tests are failing on python 3 as the new test needs to be skipped on systems without unicode support, and incorrectly passing on python 2 as they are being interpreted as bytes (need u'' prefix).

The changes to _fscodec are quite big, which will make it a pain to remove when python 2.6/2.7 is dropped. Making two new functions would reduce the patch (there are only four places which use fsencode, and two which use fsdecode), and make it easier to maintain the compat code (we can remove _fscodec on python 2.6/2.7 being dropped, and fspath on python 3.5 being dropped). Also, it'd make it clearer for new code which needs to deal with filenames (e.g. #809).

@JGoutin can you make the necessary changes, or do you want me to go ahead and do them?

JGoutin commented:
I make it for have the following easy:

To remove the `fscodec ` patch to drop Python 2.6/2.7, you only need to remove lines 43 to 93 and replace them by 
```
from os import fsencode as os_fsencode
from os import fsdecode as os_fsdecode
```

I didn't change anything to `_fscodec`, only change the try/except to define function only if needed (Previously, functions were defined everytime), and renamed them to `os_fsencode`, `os_fsdecode`

To remove `fspath `patch to drop Python 3 < 3.6, you only need to remove lines 8 to 40 and replace them by `from os import fspath`.

Code for `fscodec ` patch for Python >= 3.6 is only after line 96.


You can go ahead if you have a better idea in mind.  :wink:


Changes I did:
+ Updated with the 'u' prefix.
+ Fixed Python 3.6 CI with your PR.
+ Not skipped test for systems without Unicode support, Is there as function for detect this ?

JGoutin commented:
Updated in #841
